evm/db/journal.py

```diff
@@ -1,226 +1,264 @@
+        the changes from any subsequent changeset, giving precidence to
+        later changesets.
         """
-        idx = self.checkpoints.index(checkpoint_id)
+        if changeset_id not in self.journal_data:
+            raise KeyError("Unknown changeset: {0}".format(changeset_id))
 
-        # update the checkpoint list
-        checkpoint_ids = self.checkpoints[idx:]
-        self.checkpoints = self.checkpoints[:idx]
+        all_ids = tuple(self.journal_data.keys())
+        changeset_idx = all_ids.index(changeset_id)
+        changesets_to_pop = all_ids[changeset_idx:]
 
-        # we pull all of the checkpoints *after* the checkpoint we are
-        # reverting to and collapse them to a single set of keys that need to
-        # be reverted (giving precidence to earlier checkpoints).
-        revert_data = merge(*(
+        # we pull all of the changesets *after* the changeset we are
+        # reverting to and collapse them to a single set of keys (giving
+        # precedence to later changesets)
+        changeset_data = merge(*(
             self.journal_data.pop(c_id)
             for c_id
-            in reversed(checkpoint_ids)
+            in changesets_to_pop
         ))
 
-        return dict(revert_data.items())
+        return changeset_data
 
-    def commit_checkpoint(self, checkpoint_id):
+    def commit_changeset(self, changeset_id):
         """
-        Collapses all changes for the givent checkpoint into the previous
-        checkpoint if it exists.
+        Collapses all changes for the given changeset into the previous
+        changesets if it exists.
         """
-        changes_to_merge = self.pop_checkpoint(checkpoint_id)
-        if self.checkpoints:
-            # we only have to merge the changes into the latest checkpoint if
+        changeset_data = self.pop_changeset(changeset_id)
+        if not self.is_empty():
+            # we only have to merge the changes into the latest changeset if
             # there is one.
             self.latest = merge(
-                changes_to_merge,
                 self.latest,
+                changeset_data,
             )
+        return changeset_data
+
+    #
+    # Database API
+    #
+    def get(self, key):
+        """
+        For key lookups we need to iterate through the changesets in reverse
+        order, returning from the first one in which the key is present.
+        """
+        for changeset_data in reversed(self.journal_data.values()):
+            try:
+                value = changeset_data[key]
+            except KeyError:
+                continue
+            else:
+                if value is None:
+                    raise KeyError(key)
+                else:
+                    return value
+        else:
+            raise KeyError(key)
+
+    def set(self, key, value):
+        self.latest[key] = value
+
+    def exists(self, key):
+        for changeset_data in reversed(self.journal_data.values()):
```

Comment:
Looking at this we can de-dup some of this code by changing this implementation to be:

```python
def exists(self, key):
    try:
        self.get(key)
    except KeyError:
        return False
    else:
        return True
```